src/ExpressionParser.php

```diff
@@ -104,18 +129,37 @@ public function parseExpression($precedence = 0, $allowArrow = false)
-        if ($expr instanceof AddBinary || $expr instanceof SubBinary) {
-            /** @var AbstractExpression $left */
-            $left = $expr->getNode('left');
-            /** @var AbstractExpression $right */
-            $right = $expr->getNode('right');
-            if (
-                ($left instanceof ConcatBinary && !$left->hasExplicitParentheses())
-                ||
-                ($right instanceof ConcatBinary && !$right->hasExplicitParentheses())
-            ) {
-                trigger_deprecation('twig/twig', '3.15', \sprintf('As "+" / "-" will have a higher precedence than "~" in Twig 4.0, please add parentheses to keep the current behavior in "%s" at line %d.', $this->parser->getStream()->getSourceContext()->getName(), $token->getLine()));
+        // Check that the all nodes that are between the 2 precedences have explicit parentheses
+        if (!$expr->hasAttribute('operator') || !isset($this->precedenceChanges[$expr->getAttribute('operator')])) {
+            return;
+        }
+
+        if (str_starts_with($unaryOp = $expr->getAttribute('operator'), 'unary')) {
+            if ($expr->hasExplicitParentheses()) {
+                return;
+            }
+            $target = explode('_', $unaryOp)[1];
+            $targetPackage = $this->unaryOperators[$target]['package'];
+            $targetVersion = $this->unaryOperators[$target]['version'];
+            /** @var AbstractExpression $node */
+            $node = $expr->getNode('node');
+            foreach ($this->precedenceChanges as $operatorName => $changes) {
+                if (!in_array($unaryOp, $changes)) {
+                    continue;
+                }
+                if ($node->hasAttribute('operator') && $operatorName === $node->getAttribute('operator')) {
+                    trigger_deprecation($targetPackage, $targetVersion, \sprintf('Add explicit parentheses around the "%s" unary operator to avoid behavior change in Twig 4.0 as its precedence will change in "%s" at line %d.', $target, $this->parser->getStream()->getSourceContext()->getName(), $token->getLine()));
+                }
+            }
+        } else {
+            foreach ($this->precedenceChanges[$expr->getAttribute('operator')] as $operatorName) {
+                foreach ($expr as $node) {
+                    /** @var AbstractExpression $node */
+                    if ($node->hasAttribute('operator') && $operatorName === $node->getAttribute('operator') && !$node->hasExplicitParentheses()) {
+                        $op = explode('_', $operatorName)[1];
+                        trigger_deprecation($this->binaryOperators[$op]['package'], $this->binaryOperators[$op]['version'], \sprintf('Add explicit parentheses around the "%s" binary operator to avoid behavior change in Twig 4.0 as its precedence will change in "%s" at line %d.', $op, $this->parser->getStream()->getSourceContext()->getName(), $token->getLine()));
```

Comment:
this should say `in the next major version` as the package might not be Twig

Reply:
Indeed